src/iris/api.py

```diff
@@ -778,37 +778,39 @@ def process_resource(self, req, resp, resource, params):  # pragma: no cover
-                else:
-                    text = '%s %s %s %s' % (window, method, path, body)
-                HMAC = hmac.new(api_key, text, hashlib.sha512)
-                digest = base64.urlsafe_b64encode(HMAC.digest())
-                if equals(client_digest, digest):
-                    req.context['app'] = app
-                    if username_header:
-                        req.context['username'] = username_header
-                    return
-                else:
+                for api_key in (str(app['key']), str(app['secondary_key'])):
+                    # If username header is present, throw that into the hmac validation as well
                     if username_header:
-                        text = '%s %s %s %s %s' % (window - 1, method, path, body, username_header)
+                        text = '%s %s %s %s %s' % (window, method, path, body, username_header)
                     else:
-                        text = '%s %s %s %s' % (window - 1, method, path, body)
+                        text = '%s %s %s %s' % (window, method, path, body)
                     HMAC = hmac.new(api_key, text, hashlib.sha512)
                     digest = base64.urlsafe_b64encode(HMAC.digest())
                     if equals(client_digest, digest):
                         req.context['app'] = app
                         if username_header:
                             req.context['username'] = username_header
+                        return
                     else:
+                        # Try again with window - 1
                         if username_header:
-                            logger.warn('HMAC doesn\'t validate for app %s (passing username %s)', app['name'], username_header)
+                            text = '%s %s %s %s %s' % (window - 1, method, path, body, username_header)
                         else:
-                            logger.warn('HMAC doesn\'t validate for app %s', app['name'])
-                        raise HTTPUnauthorized('Authentication failure', '', [])
+                            text = '%s %s %s %s' % (window - 1, method, path, body)
+                        HMAC = hmac.new(api_key, text, hashlib.sha512)
+                        digest = base64.urlsafe_b64encode(HMAC.digest())
+                        if equals(client_digest, digest):
+                            req.context['app'] = app
+                            if username_header:
+                                req.context['username'] = username_header
+                                return
+                # No successful HMACs match, fail auth.
+                if username_header:
+                    logger.warn('HMAC doesn\'t validate for app %s (passing username %s)', app['name'], username_header)
+                else:
+                    logger.warn('HMAC doesn\'t validate for app %s', app['name'])
+                    raise HTTPUnauthorized('Authentication failure', '', [])
```

Comment:
I might be missing something, but if debug is turned off and no app key is correct but you do pass a username header, i'm not sure if the request will be blocked if neither for loops `return`